docs/lensvm/lens-vm.md

```diff
@@ -0,0 +1,161 @@
+| `transform() -> unsigned8` | Exported | Yes     | Core transformation logic. Pulls zero or more inputs using `next()`, applies transformation, and returns a pointer to a single output item. Supports stateful transformations. |
+| `inverse() -> unsigned8`   | Exported | No     | Optional reverse transformation logic, same interface as `transform()`. |
+
+### WASM data format
+
+Data passed across the WASM boundary—into `set_param()`, `transform()`, and `inverse()`—follows this binary format:
+
+```json
+
+[TypeId][Length][Payload]
+```
+
+- **TypeId**: Signed 8-byte integer  
+- **Length**: Unsigned 32-byte integer (optional depending on TypeId)  
+- **Payload**: Raw binary or serialized data (e.g., JSON)
+
+#### TypeId Values
+
+| TypeId | Meaning            | Notes |
+|:-------|:--------------------|:------|
+| `-1`   | Error               | May include an error message in the Payload. |
+| `0`    | Nil                | No `Length` or `Payload`. |
+| `1`    | JSON               | Payload contains a JSON-serialized object. |
+| `127`  | End of Stream      | Indicates there are no more items to process. |
+
+### Rust SDK
+
+To simplify Lens development, a [Rust SDK](https://docs.rs/lens_sdk) is available. It handles memory allocation, pointer management, and payload formatting, so you can focus on writing transformation logic.
+
+Available on [crates.io](https://crates.io/crates/lens_sdk) and GitHub: [lens-vm/lens](https://github.com/lens-vm/lens)
+
+### Example Lenses
+
+Example Lenses written in:
+
+- [Rust](https://www.rust-lang.org/)
+- [AssemblyScript](https://www.assemblyscript.org/)
+
+can be found in this repository and in [DefraDB](https://github.com/worldsibu/defradb).
+
+## Building
+
+Once you have the listed prerequisites installed, you should be able to build everything in the repository and run all the tests by running `make test` from the repository root.
+
+### Prerequisites
+
+The following tools need to be installed and added to your PATH before you can build the full contents of the repository:
+
+- [rustup](https://www.rust-lang.org/tools/install) and Cargo/rustc, typically installed via rustup.
+  - Please pay attention to any prerequisites, for example if on Ubuntu you may need to install the `build-essential` package
+- If connection errors are experienced when retrieving rust package dependencies from crates.io, you might need to tweak your `.gitconfig` as per this [comment](https://github.com/rust-lang/cargo/issues/3381#issuecomment-1193730972).
+- `npm`, typically installed via [nvm](https://github.com/nvm-sh/nvm#install--update-script)
+- [Go](https://golang.google.cn/doc/install)
+
+## Basic Lens Example
+
+The easiest way to get started writing a Lens is by using Rust, thanks to the [`lens_sdk`](https://docs.rs/lens_sdk/0.8.0/lens_sdk) crate, which provides helpful macros and utilities for Lens development.
```

Comment:
todo: I think we should avoid linking to specific versions of the lens sdk from this document, please use `https://docs.rs/lens_sdk` or `https://docs.rs/lens_sdk/latest/lens_sdk/` (`https://docs.rs/lens_sdk/latest/lens_sdk/` is required for child links, such as `https://docs.rs/lens_sdk/0.8.0/lens_sdk/macro.define.html#examples`)